File: perf-results/cross-lang/benchmarks/spectralnorm.lua
local function A(i, j)
  local ij = i + j
  return 1.0 / (ij * (ij + 1) * 0.5 + i + 1)
end

local function Av(x, y, n)
  for i = 0, n-1 do
    local val = 0
    for j = 0, n-1 do val = val + A(i, j) * x[j] end
    y[i] = val
  end
end

local function Atv(x, y, n)
  for i = 0, n-1 do
    local val = 0
    for j = 0, n-1 do val = val + A(j, i) * x[j] end
    y[i] = val
  end
end

local function AtAv(x, y, n)
  local u = {}
  for i = 0, n-1 do u[i] = 0 end
  Av(x, u, n)
  Atv(u, y, n)
end

local N = tonumber(arg and arg[1]) or 100
local u, v = {}, {}
for i = 0, N-1 do u[i] = 1.0; v[i] = 0.0 end

for _ = 1, 10 do
  AtAv(u, v, N)
  AtAv(v, u, N)
end

local vBv, vv = 0, 0
for i = 0, N-1 do
  vBv = vBv + u[i] * v[i]
  vv = vv + v[i] * v[i]
end

io.write(string.format("%.9f\n", math.sqrt(vBv / vv)))

File: perf-results/cross-lang/benchmarks/spectralnorm.java
// spectralnorm — idiomatic single-threaded Java, matched to the Python/Ruby reference
// Usage: java spectralnorm <n>

public class spectralnorm {
    static double a(int i, int j) {
        return 1.0 / ((i + j) * (i + j + 1) / 2 + i + 1);
    }

    static void av(double[] x, double[] y, int n) {
        for (int i = 0; i < n; i++) {
            double val = 0.0;
            for (int j = 0; j < n; j++) val += a(i, j) * x[j];
            y[i] = val;
        }
    }

    static void atv(double[] x, double[] y, int n) {
        for (int i = 0; i < n; i++) {
            double val = 0.0;
            for (int j = 0; j < n; j++) val += a(j, i) * x[j];
            y[i] = val;
        }
    }

    static void atav(double[] x, double[] y, int n) {
        double[] u = new double[n];
        av(x, u, n);
        atv(u, y, n);
    }

    public static void main(String[] args) {
        int n = args.length > 0 ? Integer.parseInt(args[0]) : 100;
        double[] u = new double[n];
        double[] v = new double[n];
        for (int i = 0; i < n; i++) u[i] = 1.0;

        for (int i = 0; i < 10; i++) {
            atav(u, v, n);
            atav(v, u, n);
        }

        double vBv = 0.0, vv = 0.0;
        for (int i = 0; i < n; i++) {
            vBv += u[i] * v[i];
            vv += v[i] * v[i];
        }

        System.out.printf("%.9f%n", Math.sqrt(vBv / vv));
    }
}
